File: Data Structures & Algorithms/buy-and-sell-crypto/submission-4.java
class Solution {
    public int maxProfit(int[] prices) {
        /**
            have two pointers
            move the left pointer if the right is less than left else
            move right if its greater than equal too
        **/


        int maxProfit = 0;
        int left = 0;
        for (int right = 1; right < prices.length; right++) {
            if (prices[right] > prices[left]) {
                maxProfit = Math.max(maxProfit, prices[right] - prices[left]);
            } else {
                left = right;
            }
        }

        return maxProfit;
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic = {}

        left = 0
        longest_sub = 0

        for right in range(len(s)):
            chararcter = s[right]
            if chararcter in dic:
                left = max(dic[chararcter] + 1, left)
                dic[chararcter] = right

            dic[chararcter] = right
            longest_sub = max(longest_sub, right - left + 1)

        return longest_sub
